Source/Tools/LinuxEmulation/LinuxSyscalls/SyscallsSMCTracking.cpp

```diff
@@ -201,54 +206,80 @@ void SyscallHandler::TrackMmap(FEXCore::Core::InternalThreadState* Thread, uintp
       Resource = nullptr;
     }
 
-    VMATracking.TrackVMARange(CTX, Resource, Base, Offset, Size, VMATracking::VMAFlags::fromFlags(Flags), VMATracking::VMAProt::fromProt(Prot));
+    VMATracking.TrackVMARange(CTX, Resource, Result, offset, Size, VMATracking::VMAFlags::fromFlags(flags), VMATracking::VMAProt::fromProt(prot));
   }
 
   if (SMCChecks != FEXCore::Config::CONFIG_SMC_NONE) {
     // VMATracking.Mutex can't be held while executing this, otherwise it hangs if the JIT is in the process of looking up code in the AOT JIT.
-    _SyscallHandler->TM.InvalidateGuestCodeRange(Thread, (uintptr_t)Base, Size);
+    _SyscallHandler->TM.InvalidateGuestCodeRange(Thread, Result, Size);
   }
+
+  return Result;
 }
 
-void SyscallHandler::TrackMunmap(FEXCore::Core::InternalThreadState* Thread, uintptr_t Base, uintptr_t Size) {
-  Size = FEXCore::AlignUp(Size, FEXCore::Utils::FEX_PAGE_SIZE);
+uint64_t SyscallHandler::EmulateMunmap(FEXCore::Core::InternalThreadState* Thread, void* addr, size_t length,
+                                       fextl::move_only_function<uint64_t(void* addr, size_t length)> Callback) {
+  uint64_t Result {};
+  uint64_t Size = FEXCore::AlignUp(length, FEXCore::Utils::FEX_PAGE_SIZE);
 
   {
     // Frontend calls this with nullptr Thread during initialization.
     // This is why `GuardSignalDeferringSectionWithFallback` is used here.
     // To be more optimal the frontend should provide this code with a valid Thread object earlier.
     auto lk = FEXCore::GuardSignalDeferringSectionWithFallback(VMATracking.Mutex, Thread);
+    Result = Callback(addr, length);
+    if (FEX::HLE::HasSyscallError(Result)) {
+      return Result;
+    }
 
-    VMATracking.DeleteVMARange(CTX, Base, Size);
+    VMATracking.DeleteVMARange(CTX, reinterpret_cast<uintptr_t>(addr), Size);
   }
 
   if (SMCChecks != FEXCore::Config::CONFIG_SMC_NONE) {
-    _SyscallHandler->TM.InvalidateGuestCodeRange(Thread, (uintptr_t)Base, Size);
+    _SyscallHandler->TM.InvalidateGuestCodeRange(Thread, reinterpret_cast<uintptr_t>(addr), Size);
```

Comment:
For personal reference, is this affected by the same race condition? A sequence of munmap followed by mmap could be scheduled in such a way that we call InvalidateGuestCodeRange after having mapped the code.

This is probably fine since it would additionally require code to be compiled before InvalidateGuestCodeRange gets called (leaving a very slim time window), and even then the worst that'll happen is probably a recompile. But with code caching, the effect would be that a cache that we just loaded gets discarded immediately, so I should double-check this interaction.

Reply:
Currently this could just cause some additional invalidation (that would have already occurred), plus there's some nasty interactions between recursive mutex locking that can't be solved, which is why this mutex juggle needs to occur like this.